Initial version of the code-signing dataplane service

Jaxelr opened the pull request:
# Data Plane API - Pull Request
Initial onboarding definition for the dataplane service of the Azure Code Signing service.
See: https://github.com/Azure/cadl-azure/issues/1999
cc: @DawnWang17 

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/23716
fix https://github.com/azure-sdk/TestingForEngagementExperience/issues/296
fix https://github.com/azure-sdk/TestingForEngagementExperience/issues/297

Jaxelr committed:
Initial version of the code-signing dataplane service

Jaxelr committed:
Remove error objects since those come from the Azure Core namespace

Jaxelr committed:
fix: added azure.codesigning.json resulting file

Jaxelr committed:
fix: use newer version dependency syntax

Jaxelr committed:
Rolling back version change per PR feedback

Jaxelr committed:
fix: polling operation mapping per PR feedback

Jaxelr committed:
Adding x-ms-examples for example validation

Jaxelr commented on specification/codesigning/Microsoft.CodeSigning/main.cadl:
```diff
@@ -0,0 +1,158 @@
+
+  @summary("Initiates a sign operation.")
+  @doc("Initiates a sign operation.")
+  @pollingOperation("GetSignOperation")
```
Thanks, this is fixed on commit 384fd05

Jaxelr committed:
fix: added custom-words per PR feedback

Jaxelr committed:
Renamed parameter definition to match open api json definition

ruowan commented:
/azp run

Jaxelr commented:
@msftbot dont close, still had work scheduled.

Jaxelr committed:
fix: modified the api-version to synch up with the folder api-version

Jaxelr committed:
Adding details json cadl validation

Jaxelr committed:
refactor: moved from cadl to typespec

Jaxelr committed:
chore: attach azure.codesiging.json definition

Jaxelr committed:
chore: remove package-lock.json since its unneeded.

Jaxelr committed:
feat: adding get sign eku operation & example

fix: sign long running operation

Jaxelr committed:
chore: removed unused swagger.json generated

Jaxelr committed:
chore: removing data-plane swagger generated, keeping just the typespec project

Jaxelr committed:
Readding the swagger information for dataplane service

Jaxelr committed:
fix: ci issues pointed

examples are not referenced on the swagger file
resources should not be read only

Jaxelr committed:
added examples to typespec definition

Jaxelr committed:
fix: examples

Jaxelr committed:
more fixes for example.json operations

Jaxelr committed:
even more fixes for example.json operations

Jaxelr commented:
/azp run

Jaxelr committed:
fix: minor

Remove optional parameter from region. since its required.
Added description to namespace definition
Removed optional parameter from operation status, since the id is required.
Added suppression of warning

Jaxelr committed:
Added fixes to azure.codesigning.json generation

Jaxelr committed:
Added required property of id to examples

Jaxelr committed:
Switched property to a more descriptive naming for the eku response

Jaxelr committed:
fix: moved examples to a non version specific folder

Jaxelr committed:
fix: the compiler didnt liked not using the preview folder, rolling back

mikekistler reviewed (changes requested):
This looks very good but there are few minor problems. Most should be easy to fix.

Jaxelr committed:
feat: added oauth2 security definition

Jaxelr committed:
fix: removed api-version from example path

Jaxelr committed:
style: change casing of operations to camel case

fix: per guidance modified the getsignEku operation into listSignEku

Jaxelr committed:
Update azure.codesigning.json generated by tsp compiler

Jaxelr commented on specification/codesigning/Microsoft.CodeSigning/main.tsp:
```diff
@@ -0,0 +1,169 @@
+}
+
+interface CertificateProfileOperations {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is a custom operation status endpoint."
```
The reason was that I needed a custom route with this path:

/codesigningaccounts/{codeSigningAccountName}/certificateprofiles/{certificateProfileName}/**sign**/{operationId}

and the GetResourceOperationStatus appends an operations path inside the path:

"/codesigningaccounts/{codeSigningAccountName}/certificateprofiles/{certificateProfileName}/**operations**/{operationId}"

Jaxelr committed:
fix: modified ResourceAction to LongRunningResourceAction per feedback

Jaxelr commented on specification/codesigning/Microsoft.CodeSigning/main.tsp:
```diff
@@ -0,0 +1,169 @@
+  @doc("Initiates a sign operation.")
+  @example("./examples/2022-06-15-preview/CertificateProfile_Sign.json", "Sign")
+  @pollingOperation(CertificateProfileOperations.GetSignOperation)
+  Sign is ResourceAction<CertificateProfile, SignRequest, OperationStatus>;
```
I agree. Committed the modification.

Jaxelr committed:
docs: added enum descriptions to each value on hash algorithms

mikekistler reviewed (approved):
Thanks for the fixes! This is looking really good now. I think it is ready to bring to the API Stewardship review whenever you are ready.

Jaxelr committed:
fix: pinned to current versions to avoid ci errors by using latest dependency

Jaxelr committed:
Added prefix of StandardResourceOperations interface to Resource operations

Jaxelr committed:
fix: sign operation should return 202 Accepted on example

Jaxelr committed:
fix: sign operation should return 202 Accepted on example

Jaxelr committed:
style: tsp format result

Jaxelr committed:
refactor: removed unused mapping to parent resource

Jaxelr committed:
refactor: removing typespec-python emitter

Jaxelr committed:
Adding get sign root certificate operation

Jaxelr committed:
fix: adding custom word rootcert

Jaxelr committed:
fix: adding api version parameter

Jaxelr committed:
Added resource provider folder per ci finding

weshaggard commented on specification/codesigning/Microsoft.CodeSigning/tspconfig.yaml:
```diff
@@ -0,0 +1,10 @@
+emit:
+  - "@azure-tools/typespec-autorest"
+options:
+  "@azure-tools/typespec-autorest":
```
I assume you want to generate SDK's and if so, you will want add the other sdk configurations. See our sample at https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml.

Jaxelr replied:
I used the sample for now to add those emitters. Thanks

weshaggard commented on specification/codesigning/Microsoft.CodeSigning/package.json:
```diff
@@ -0,0 +1,14 @@
+{
```
You might also need to update your branch with the latest changes in main to pull in the newer version of the package.json in the root of the repo.

Jaxelr replied:
This is done, thanks!

weshaggard commented on specification/codesigning/Microsoft.CodeSigning/main.tsp:
```diff
@@ -0,0 +1,221 @@
+import "@typespec/rest";
```
Let please remove "Microsoft." from the path as we try to keep those redundant names out of the paths. See https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md for some of the structure guidelines.

Jaxelr replied:
Thank you, addressed feedback.

Jaxelr committed:
chore: removing package.json per PR feedback

Jaxelr committed:
Renaming folder to remove "Microsoft." path per feedback from PR

see: https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md for guidelines

Jaxelr committed:
feat: expanded the typespec options to include the lang generation

mikekistler commented on specification/codesigning/CodeSigning/main.tsp:
```diff
@@ -0,0 +1,221 @@
+}
+
+interface CertificateProfileOperations {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is a custom operation status endpoint."
```
Why can you not use an operation template from Azure Core or Foundations? It looks like Foundations.GetOperationStatus should work.

mikekistler replied:
Sorry ... I see now that you are using GetOperationStatus. Perhaps we need to promote this into Azure Core proper.

Jaxelr replied:
My assumption is that there's no action needed from my end on this finding, as of now. I'm marking as resolved, but feel free to unresolve otherwise.

mikekistler commented on specification/codesigning/CodeSigning/main.tsp:
```diff
@@ -0,0 +1,221 @@
+    OperationStatus
+  >;
+
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is a custom operation status endpoint."
```
The explanation for this suppression looks wrong -- looks like it was just copied from above. Please consider using an operation template or provide a better explanation here.

If you do not change to use an operation template, you must add an api-version parameter and a default (error) response.

Jaxelr replied:
This will be resolved using the suggestion below.

mikekistler reviewed (changes requested):
Please use operation templates from Azure Core wherever possible. If not possible, please provide a clear explanation.

markweitzel reviewed (approved):
Per #23716 -- API Stewardship Board approved for Preview

Jaxelr committed:
Fix: Based on review from API stewardship team

Added more detailed information for docs and summary of operations
Refactored getSignRootCertificate operation to use RpcOperation instead of custom operation
Pluralized the listSignEkus operation
Renamed models to provide more detailed description of their purposed
Eliminated the custom enum in favor of the typespec enum provided

Jaxelr committed:
refactor: moved version to 2023-06-15 based on feedback of ACS Eng team

Jaxelr committed:
refactor: renamed models following azure team feedback

raych1 committed:
Update tspconfig.yaml

Applied latest schema of emitter options and parameters.

Jaxelr commented:
@msftbot dont close, I've been on holidays and will pick up this work on this sprint.

weshaggard commented on custom-words.txt:
```diff
@@ -493,9 +494,8 @@ codegen
-codespaces
-coffeelake
+codesigning
+Coffeelake
```
Why did you remove the codespace entries? Also please use lowercase.

Jaxelr replied:
I must've messed up the rebase. Thanks for the catch, this is fixed.

Jaxelr committed:
fix: adding back missing custom words, fixing coffeelake casing

Jaxelr committed:
docs: updated samples to include varied hashes to represent a more familiar request for customers

Jaxelr committed:
docs: updated the docs clauses of the typespec file to avoid repetitions on the docs generated.

Jaxelr committed:
fix: update README.md of data-plane generation to include the correct tag

Jaxelr committed:
Regenerated azure.codesigning.json file

Jaxelr committed:
Based on pr build feedback, updated custom words definition

Jaxelr committed:
docs: per compiler feedback, adding doc definition for versioning

Jaxelr committed:
fix: outdated examples of data-plane swagger

mikekistler reviewed (changes requested):
One more small suggested change.

Jaxelr committed:
fix: per PR feedback, using T payload for Foundations.OperationStatus

mikekistler reviewed (approved):
Looks good! 👍

Jaxelr commented:
@weshaggard / @lmazuel anything else missing from my end?

Jaxelr committed:
chore: updated Azure.CodeSigning swagger generated using tsp v0.46.0

mikeharder commented on specification/codesigning/CodeSigning/main.tsp:
```diff
@@ -0,0 +1,207 @@
+import "@typespec/rest";
+import "@typespec/versioning";
+import "@azure-tools/typespec-azure-core";
+import "@azure-tools/typespec-autorest";
```
```suggestion
```

Your spec should not requires this import.

Jaxelr replied:
Fixed. Thanks

mikeharder reviewed (dismissed):
TypeSpec Validation is failing

Jaxelr committed:
fix: removed autorest from the dependencies per feedback

fix: added suppression for custom RPC operation

Jaxelr committed:
Removing json examples for compliance with avocado pipeline

Jaxelr committed:
fix: Adding examples to tsp directory

Jaxelr committed:
Corrected samples

lmazuel commented:
Force merging the PR, as the broken CI are caused by known bugs:
- https://github.com/Azure/typespec-azure/issues/1674
- https://github.com/Azure/typespec-azure/issues/3349

weidongxu-microsoft commented:
Note that the example of the LRO uses `azure-async-operation`. Therefore, either the example is wrong, or the LRO is wrong.

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/codesigning/data-plane/Azure.CodeSigning/preview/2023-06-15-preview/examples/CertificateProfileOperations_Sign.json#L25C10-L25C31

Jaxelr commented:
I can submit a PR to fix it, thanks for catching this up.

weidongxu-microsoft commented:
Thanks @Jaxelr 

If possible, would you also give us an example of the last poll (when success)?

As LongRunningResourceAction, the final poll should be as shape
```
{
  "id": "##",
  "status": "succeeded",
  "result": {
    ...
  }
}
```

Jaxelr commented:
I can definitely do that, Thanks.

weidongxu-microsoft commented:
Thanks. Related doc on LRO of LongRunningResourceAction
https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md#long-running-action-operations
